platform-hub-api/app/models/audit.rb

```diff
@@ -0,0 +1,67 @@
+class Audit < ApplicationRecord
+
+  belongs_to :auditable, polymorphic: true
+  belongs_to :associated, polymorphic: true
+  belongs_to :user
+
+  before_create :set_associated_if_needed
+  before_create :set_descriptors_if_needed
+  before_create :denormalise_user_fields
+  after_create :output_log_message
+  after_find { readonly! }
+  before_destroy { raise ActiveRecord::ReadOnlyRecord }
+
+  scope :ascending, -> { reorder(version: :asc) }
+  scope :descending, -> { reorder(version: :desc) }
+  scope :by_action, -> (action) { where(action: action) }
+  scope :by_auditable, -> (auditable_id, auditable_type) { where(auditable_id: auditable_id, auditable_type: auditable_type) }
+  scope :by_associated, -> (associated_id, associated_type) { where(associated_id: associated_id, associated_type: associated_type) }
+
+  def message
+    m = []
+    m << "[Audit] Action: #{self.action}"
+    m << "By user: #{self.user.id} - #{self.user_name} - #{self.user_email}" if self.user
+    m << "On thing: #{self.auditable_type} - #{self.auditable_id} - #{self.auditable_descriptor}" if self.auditable
+    m << "Associated to: #{self.associated_type} - #{self.associated_id} - #{self.associated_descriptor}" if self.associated
+    m << "Comment: #{self.comment}" if self.comment
+    m << "Remote IP: #{self.remote_ip}" if self.remote_ip
+    m << "Request UUID: #{self.request_uuid}" if self.request_uuid
+    m.join(' | ')
```

Comment:
Agreed… but, later :) Right now we don't have JSON output logging set up in the API stack, but when we do, we'll need to review all log messages and we can change this then. Does that sound okay to you? Or do you think we should do this now? (I'm thinking that right now it's better to have something more bespoke and descriptive like this in the logs, but later we can just output the whole `Audit` entry as a JSON object).